### DISCUSSION TITLE
When to use contains_eager, joinedload and .execution_options(populate_existing=True)

### DISCUSSION THREAD

**Discussion by OddBlueDog** [sqlalchemy/sqlalchemy#7654] [Usage Questions] [answered]
Is this the correct way to use contains_eager to load in the 'Client' and then the clients publisher and author relationships. Should they both use .contains_eager or is it OK to use joinedload to avoid having to use outerjoin on each?

I only want the Client.book to contain the filtering applied to the explicit outerjoin.

Another question, should I always use `.execution_options(populate_existing=True)` if filtering and using contains_eager?

```
class Client(Base):
    __tablename__ = "client"

    id = Column(Integer, primary_key=True, index=True)
    name = Column(String,  nullable=False)
    company_name = Column(String)
    is_active = Column(Boolean(), default=True)

    # Forgien keys
    organisation_id = Column(Integer, ForeignKey("organisation.id", ondelete="CASCADE"), nullable=False,  index=True)

    # relationships
    book = relationship("Book", back_populates="client")
  ```

```
class Book(Base):
    __tablename__ = "book"

    id = Column(Integer, primary_key=True, index=True)
    narrative = Column(String, index=True, nullable=False)
    publish_date= Column(DATE, index=True,  nullable=False, server_default=func.now())
    chapters= Column(SmallInteger, default=PaymentStatus.pending.value)

    # Forgien Keys
    author_id= Column(Integer, ForeignKey("author.id"), nullable=False, index=True)
    publisher_id= Column(Integer, ForeignKey("publisher.id"), nullable=False, index=True)
    client_id = Column(Integer, ForeignKey("client.id", ondelete="CASCADE"), nullable=False, index=True)

    # Relationships
    client = relationship("Client", back_populates="book")
    publisher = relationship("Publisher", back_populates="book")
    author = relationship("Author", back_populates="book")
```

Is this OK?
```       
query = select(Client).outerjoin(Book, and_(Book.client_id == Client.id, Book.publish_date<= last_day_of_last_month)) \
                .where(Client.organisation_id == organisation_id)\
                .options(contains_eager(Client.book)) \
                .options(contains_eager(Client.book).joinedload(Book.author)) \
                .options(contains_eager(Client.book).joinedload(Book.publisher)) \
                .execution_options(populate_existing=True)
```

or should I do this with contains eager and explicit outerjoins?

```       
query = select(Client).outerjoin(Book, and_(Book.client_id == Client.id, Book.publish_date<= last_day_of_last_month)) \
                .where(Client.organisation_id == organisation_id)\
                .outerjoin(Book.publisher) \
                .outerjoin(Book.author) \
                .options(contains_eager(Client.book)) \
                .options(contains_eager(Client.book).contains_eager(Book.author)) \
                .options(contains_eager(Client.book).contains_eager(Book.publisher)) \
                .execution_options(populate_existing=True)
```

**Comment by zzzeek** [comment, marked as answer]
I think once you're in contains_eager() you are now responsible for writing out the joins, so the latter, contains_eager() all the way with explicit outerjoin() all the way.   you can alternatively use other kinds of loading such as selectinload() and then you wouldn't have to do it this way.   also w/ contains_eager() , populate_existing is usually what people want, yes.

**Reply by OddBlueDog** [reply]
Thanks for your super quick answer.

I think the queries looked identical between the two, but wasn’t sure if it would cause any issues with sqlalchemy / the session using joinedload to make it more succinct or if it was a bit “hacky”. I will stick to using contains_eager all the way down.

I’ve been struggling to choose between selectinload and joinedload. Is there a case where one is more preferred to the other for instance if it’s one to one, one to many or many to many?

**Reply by CaselIT** [reply]
One on one are ideal for joined load, since no data needs to be discarded.
In general if there are few elements in the relationship, or the parent table as few small columns it's a wash.

For a relationship that can have many elements (say 100+) in general selectin is probably better.

Another thing to consider is the network latency. If it's high then joined may be preferable since it reduces the number if queries

As always the real answer is "it depends". Do some testing to select what is more appropriate in your case
